src/service/SongsService.tsx

```diff
@@ -50,26 +47,106 @@ export async function listSongs(searchString: string): Promise<Song[]> {
   } else if (isNumeric(searchString)) {
     return songs.filter((song) => song.songNumber.toString().startsWith(searchString));
   } else {
+    // Build a cache to speed up duplicate calls
+    const matchScores: Map<number, number> = new Map<number, number>();
+    songs.forEach((song) => {
+      matchScores.set(song.songNumber, getMatchScore(song, searchString));
```

Comment:
I like this cache. Do you think it'd be possible to go further with it? Like we also have a global cache `const matchScoresBySearchString: Map<string, Map<number, number>>` and then do `matchScoresBySearchString.set(searchString, matchScores` whenever we get a thing?

If it's just browser memory then it might not be that helpful since whenever they refresh the page it'd get wiped.

Actually, `local Storage` could be an option, it's basically a persistent hashmap. We would do something like `storeItem.add(searchString, matchScores.toString())` and then grab that in the future for repeated searches.

If we go with that idea though I think we will need a way to invalidate people's caches remotely, with like a version number or something in case we make an update to the algorithm.

Anyways just kinda throwing ideas around, I don't think we need to implement that this PR. This works well as is, and feels decently fast.

Reply:
Unfortunately the matchScores depend on the search string, so unless we expect people to be searching the same thing over and over, caching that might not be the best. We can however, cache the trimmed and tokenized tokens for each song so we don't have to redo that each time. I can try that.